src/Image.cc

```diff
@@ -799,8 +800,26 @@ void Image::jpegToARGB(jpeg_decompress_struct* args, uint8_t* data, uint8_t* src
 cairo_status_t
 Image::decodeJPEGIntoSurface(jpeg_decompress_struct *args) {
   cairo_status_t status = CAIRO_STATUS_SUCCESS;
-  
-  uint8_t *data = new uint8_t[naturalWidth * naturalHeight * 4];
+
+  size_t stride = naturalWidth * 4;
+  if (naturalWidth != 0 && stride / naturalWidth != 4) {
+    // The integer overflowed, which means we cannot allocate that much memory.
+    jpeg_abort_decompress(args);
+    jpeg_destroy_decompress(args);
+    this->errorInfo.set(NULL, "malloc", errno);
+    return CAIRO_STATUS_NO_MEMORY;
+  }
+
+  size_t size = naturalHeight * stride;
+  if (naturalHeight != 0 && size / naturalHeight != stride) {
+    // The integer overflowed, which means we cannot allocate that much memory.
+    jpeg_abort_decompress(args);
+    jpeg_destroy_decompress(args);
+    this->errorInfo.set(NULL, "malloc", errno);
+    return CAIRO_STATUS_NO_MEMORY;
+  }
```

Comment:
You could consider combining these two `if` bodies into one since you don't have to check overflow *immediately* after doing the mul.